engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/transformation/GenericNodeTransformation.scala

```diff
@@ -0,0 +1,51 @@
+package pl.touk.nussknacker.engine.api.context.transformation
+
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.NodeId
+import pl.touk.nussknacker.engine.api.context.{ProcessCompilationError, ValidationContext}
+import pl.touk.nussknacker.engine.api.definition.{NodeDependency, Parameter}
+/*
+  This trait provided most generic way of defining Node. In particular, implementations can dynamically define parameter list
+  and target validation context, based on current parameters.
+  Please @see LastVariableFilterTransformer for sample usage
+
+  NOTE: this is *experimental* API, subject to changes. In particular:
+   - parameters changes are *not* fully supported on UI,
+   - the trait can *only* be used for custom nodes for now (not for sources, sinks, ...)
+   - branches/joins are not supported
+   - handling dependencies probably will change
+ */
+trait GenericNodeTransformation[T] {
+
+  //ValidationContext for single input, Map[String, ValidationContext] for joins
+  type InputContext
+
+  //State is arbitrary data that can be passed between steps of NodeTransformationDefinition
+  type State
+
+  //TODO: what if we cannot determine parameters/context? With some "fatal validation error"?
+  type NodeTransformationDefinition = PartialFunction[TransformationStep, TransformationStepResult]
+
+  def contextTransformation(context: InputContext, dependencies: List[NodeDependencyValue])(implicit nodeId: NodeId): NodeTransformationDefinition
```

Comment:
WDYT about adding `Future` to returned type?  We have in a quite near future example when we would like to call external service to determine parameter's definitions.

Reply:
I thought about it, but I don't think it's a good idea. Because we'd either have to:
a) put Await somewhere nearby
b) pull Future along all compilation process

b) would be very difficult and a) wouldn't bring much benefit. I think compilation process is rather "rare" and can be considered synchronous. The schema registry case also shows that probably caching will also be involved anyway.